Update the reviewing chapter (#1320)
alan-turing-institute/the-turing-way: book/website/reproducible-research/reviewing/reviewing-checklist.md

(rr-checklist-for-code-review)=
# Checklist for code review process

This section presents some checklists for both the coder and the reviewer, as part of a formal review process.
The reviewer checklists are split into two categories: one for the whole program, and one for individual files or proposed changes.

The lists are created with a focus on good software engineering practice and are intended to be a source of inspiration.
When assessing the checklists, it is recommended to consider to what extent the item mentioned is implemented.
Some items on the lists may not apply to your project or programming language, in which case they should be disregarded.

In all cases, the goal is to use your programming experience to figure out how to make the code better.

## For the coder

- Does the new code meets the required standards of the project?
  The standards are typically written under `contributing guidelines` by the project you are contributing to.

> - Is there [documentation](#documentation) to the required standards of the project?

Reviewer: ```suggestion
- Is there [documentation](#documentation) that meets the required standards of the project?
```

- Are you following any declared {ref}`style guide<rr-code-quality>` for the project?
- Are there new [tests](#tests) for the new material, based on the required standards of the project?
  - Do these tests pass locally?
  - Are the tests in the rest of the code base still passing locally?
- Create the pull request.

> - Many {ref}`continuous integration (CI)<rr-ci>` systems will check if the tests in the main project pass automatically
> once you create a pull request. If the repository is using a CI, make sure all builds and tests complete.
> Consult the CI reports to see if your code is causing the tests in the main project to fail.

Reviewer: ```suggestion
- Many {ref}`continuous integration (CI)<rr-ci>` systems will check if the tests in the main project pass automatically once you create a pull request.
  If the repository is using a CI, make sure all builds and tests complete.
  Consult the CI reports to see if your code is causing the tests in the main project to fail.
```

- If necessary, now formally request a review.

## For the reviewer

- Check the required standards of the project. The standards are typically written under
`contributing guidelines` by the project you are contributing to.
- Check the code meets basic project {ref}`style guide<rr-code-quality>`, if this is not automatically checked by {ref}`continuous integration (CI)<rr-ci>`.

> - Are the [tests](#tests) and [documentation](#documentation) conform to the standards.

Reviewer: ```suggestion
- Do the [tests](#tests) and [documentation](#documentation) conform to the standards?
```

- Is all the code easily understood? Depending on the language, files may contain interfaces, classes or other type definitions, and functions (see [Architecture](#architecture)).
    The essential architectural concepts can be reviewed as follows:
  - Check the [interfaces](#interfaces) lists.
  - Check the [classes and types](#classes-and-types) lists.
  - Check the [function/method declarations](#function-method-declarations) lists.
  - Check the [function/method definitions](#function-method-definitions) lists.
- Do the [tests](#tests) actually ensure the code is robust in its intended use?
  - Are there any bugs or other defects?
- Are [security](#security) issues handled correctly?
  - Check the [security of new codes](#security-of-new-codes).

> - Does the new code meets the [legal requirements](#legal)?

Reviewer: ```suggestion
- Does the new code meet the [legal requirements](#legal)?
```

- Are these modules as small as they can be?
- Is there a clear, hierarchical or layered, dependency structure between
  these modules?
  - If not, the functionality should be rearranged, or perhaps heavily
    interdependent modules should be combined.
- Can the design be simplified?

### Security

If you're making software that is accessible to the outside world (for example a web
application), then security becomes important. Security issues are defects,
but not all defects are security issues. A security-conscious design can help
mitigate the security impact of defects.

- Which modules deal with user input?
- Which modules generate output?
- Are input and output compartmentalized?
  - If not, consider making separate modules that manage all input
    and output, so validation can happen in one place.
- In which modules is untrusted data present?
  - The fewer the better.
- Is untrusted data compartmentalized?
  - Ideally, validate in the input module and pass only
    validated data to other parts.

### Legal

As a developer, you should pay attention to the legal rights of the
creators of the code you're using. Here are some things to check. When in

> doubt, ask your licensing person for advice.

Reviewer: Licensing advisor?

> doubt, ask your licensing person for advice.

Reviewer 2: ```suggestion
doubt, ask someone experienced in licensing for advice.
```

(rr-checklist-for-code-review)=
## Checklist for code review process

This section presents some checklists for both the coder and the reviewer, as part of a formal review process.
The reviewer checklists are split into two categories: one for the whole program, and one for individual files or proposed changes.

The lists are created with a focus on good software engineering practice and are intended to be a source of inspiration.
When assessing the checklists, it is recommended to consider to what extent the item mentioned is implemented.
Some items on the lists may not apply to your project or programming language, in which case they should be disregarded.

In all cases, the goal is to use your programming experience to figure out how to make the code better.

## Table of contents

- [For the coder](#For_the_coder)
- [For the reviewer](#For_the_reviewer)
- [Program level checklist](#Program_level_checklist)
- [File/Change level checklist](#File/Change_level_checklist)

> <a name="For_the_coder"></a>

Reviewer 3: Naive question. Is this `<a name=` HTML tag recognized for cross-linking by the book assembly software?

Author: Good question. The[ handbook style guide](https://the-turing-way.netlify.app/community-handbook/style/style-crossref.html) does not mention something about the HTML headings. I am wondering why the inline HTML has been used in several sections of the book. My markdown linter recommends removing inline HTML. So I removed those from files in the code review section.

### For the coder

- [ ] Does the new code meets the required standards of the project?
  The standards are typically written under `contributing guidelines` by the project you are contributing to.
- [ ] Is there [documentation](#documentation) to the required standards of the project?
- [ ] Are you following any declared {ref}`style guide<rr-code-quality>` for the project?
- [ ] Are there new [tests](#tests) for the new material, based on the required standards of the project?
  - [ ] Do these tests pass locally?
  - [ ] Are the tests in the rest of the code base still passing locally?
- [ ] Create the pull request; wait for any {ref}`continuous integration (CI)<rr-ci>` checks to complete.

> - [ ] Consult the CI reports. Did all builds and tests complete? your code is not causing the tests in the main project to fail. Many CI systems will test this automatically for you once you create the PR.

Reviewer 3: CI reports can come both from the repository system in used (for instance, GitHub), or custom ones, set up at the beginning by the repository managers. But it can be overkill for beginners managing a software project

Author: The item in the checklist is now revised. Does it address your concern?

- [ ] Is the project license compatible with all libraries?

<a name="File/Change level checklist"></a>
### File/Change level checklist

When you're checking individual changes or files in a pull request, the
code itself becomes the subject of scrutiny. Depending on the language, files
may contain interfaces, classes or other type definitions, and functions. All
these should be checked.

#### Interfaces

- [ ] Is the interface documented?
- [ ] Does the concept it models make sense?
- [ ] Can it be split up further? (Interfaces should be as small as possible)

Note that most of the following items assume an object-oriented programming
style, which may not be relevant to the code you're looking at.

#### Classes and types

- [ ] Is the class documented?
- [ ] Does it have a single responsibility? Can it be split?
- [ ] If it's designed to be extended, can it be?
- [ ] If it's not designed to be extended, is it protected against that?
- [ ] If it's derived from another class, can you substitute an object of this class for one of its parent class(es)?
- [ ] Is the class testable?
  - [ ] Are the dependencies clear and explicit?
  - [ ] Does it have a small number of dependencies?

>   - [ ] Does it depend on interfaces, rather than on classes?

Reviewer 3: External programs needed by classes (for instance command-line tools called from the classes), should also be documented, as they might be optional for several usages. The installation procedure of these external programs should be documented in the installation documentation.

Author: In section **Classes and types**, Under the item " Is the class documented", an extra item is added for "External programs needed by classes". Regarding the documentation of the installation procedure of these external programs, I think it is already covered by the item "Is the use and function of third-party libraries documented?" in section **Documentation**. Is it OK?

Reviewer 3: Yes
